### PR TITLE
prevent bind exceptions from crashing Node.js

### PR DESCRIPTION
**Pull request by mnakama**
This prevents node crashes from Client.bind() and etc.
We set a callback handler in the instance during bind()
and send errors to it.

Fixes https://github.com/mcavage/node-ldapjs/issues/356

### CODE REVIEW

**Comment by agInteg** [comment]
This fixes the bind bug where there is a problem with the LDAP server is unavailable and it results in an exception that cannot be handled in an easy way. Please merge this.

**Comment by trulore** [comment]
I took the code from this PR and manually merged it into my own code base, and it fixed a critical bug we were having that crashed node several times per day.    Thank you very much for this!

**Comment by mnakama** [comment]
@trulore 
Cool, I'm glad it helped you =). It's a small change, but it took me a few hours to figure out what spot needed the change. The whole error handling system in this code is done with event dispatches... which weren't getting handled gracefully. Seems hard to follow, but this is ES5 code.

@agInteg 
Seems this project needs a new maintainer. It's been half a year since I opened this PR. Though a few people have forked my code for this fix, I'm not a good candidate to take over since I don't work at the company where I implemented this anymore (and thus I currently have no use for the code).

@mcavage 
If you still check your github notifications, maybe you'll see this. Would you mind merging this PR for the community? It's very small and fixes a crucial crash bug.

**Comment by zont** [comment]
Need this bugfix!

**Comment by sebaplaza** [comment]
I confirm, this fix is working for #356.
Please Merge, we're having serious issues with this bug.

File: lib/client/client.js [hunk 1]
```diff
@@ -362,6 +366,17 @@ function Client(options) {
 util.inherits(Client, EventEmitter);
 module.exports = Client;
 
+/**
+ * Default handler for error callbacks when
+ * one isn't set in the instance
+ */
+Client.prototype.cb = function(err, ret) {
+  if (err) {
+    console.error('Caught exception:', err);
```
Comment (jsumners) [inline lib/client/client.js]:
```suggestion
    this.log.error('Caught exception:', err);
```

File: lib/client/client.js [hunk 2]
```diff
@@ -362,6 +366,17 @@ function Client(options) {
...
+  } else {
+    console.log('Unhandled output:', ret);
```
Comment (jsumners) [inline lib/client/client.js]:
```suggestion
    this.log.info('Unhandled output:', ret);
```

**Comment by DavidIHunt** [comment]
PLEASE merge this...
It's been 2 years!

The bug it fixes is making this module awful to use. Imagine trying to call a function of a library which can asynchronously throw an error that you can't catch and which will crash your process.

**Comment by UziTech** [comment]
@DavidIHunt if you want to take this on that would be great. I think this needs to be rebased on the `next` branch and a few tests added.

**Comment by UziTech** [comment]
Is the issue here that the `this._send` call throws an error instead of passing back to the `callback`?

**Comment by mnakama** [comment]
> Is the issue here that the `this._send` call throws an error instead of passing back to the `callback`?

If I remember correctly, there was no handler for this error that I could find, so it just threw exceptions.

**Comment by UziTech** [comment]
@DavidIHunt Would this code work to catch the error?

```js
const client = ldapjs.createClient(...)
client.on('error', (err) => {
	if (err.code === 'ECONNREFUSED') {
		// handle connection refused (LDAP server not started)
	}
}).on('connectError', (err) => {
	// handle connection timeout (firewall problem)
})
client.bind(...)
```

**Comment by vdechef** [comment]
> @DavidIHunt Would this code work to catch the error?
> 
> ```js
> const client = ldapjs.createClient(...)
> client.on('error', (err) => {
> 	if (err.code === 'ECONNREFUSED') {
> 		// handle connection refused (LDAP server not started)
> 	}
> }).on('connectError', (err) => {
> 	// handle connection timeout (firewall problem)
> })
> ```

This workaround does not fix the problem when the LDAP server does not answer (for example when your account is locked).

But I did try with the _next_ package (_npm install ldapjs@next_) and I was unable to reproduce the bug. So for me it seems fixed in _next_ branch.

**Comment by jsumners** [comment]
Given the response https://github.com/ldapjs/node-ldapjs/pull/433#issuecomment-577263109

and the lack of response from original authors, I am closing this PR. If someone discovers it is _not_ resolved in the `next` branch, they are welcome to open a new PR targeting that branch which fixes the issue.

**Comment by jsumners** [comment]
⚠️ This issue has been locked due to age. If you have encountered a recent
problem that seems to be covered by this issue, please open a new issue.

Please include a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)
when opening a new issue.
